shared/archive_unix.go

```diff
@@ -0,0 +1,122 @@
+	// bz2 - 2 bytes, 'BZ' signature/magic number
+	// gz - 2 bytes, 0x1f 0x8b
+	// lzma - 6 bytes, { [0x000, 0xE0], '7', 'z', 'X', 'Z', 0x00 } -
+	// xy - 6 bytes,  header format { 0xFD, '7', 'z', 'X', 'Z', 0x00 }
+	// tar - 263 bytes, trying to get ustar from 257 - 262
+	header := make([]byte, 263)
+	_, err = f.Read(header)
+	if err != nil {
+		return []string{""}, "", err
+	}
+
+	switch {
+	case bytes.Equal(header[0:2], []byte{'B', 'Z'}):
+		return []string{"-jxf"}, ".tar.bz2", nil
+	case bytes.Equal(header[0:2], []byte{0x1f, 0x8b}):
+		return []string{"-zxf"}, ".tar.gz", nil
+	case (bytes.Equal(header[1:5], []byte{'7', 'z', 'X', 'Z'}) && header[0] == 0xFD):
+		return []string{"-Jxf"}, ".tar.xz", nil
+	case (bytes.Equal(header[1:5], []byte{'7', 'z', 'X', 'Z'}) && header[0] != 0xFD):
+		return []string{"--lzma", "-xf"}, ".tar.lzma", nil
+	case bytes.Equal(header[0:3], []byte{0x5d, 0x00, 0x00}):
+		return []string{"--lzma", "-xf"}, ".tar.lzma", nil
+	case bytes.Equal(header[257:262], []byte{'u', 's', 't', 'a', 'r'}):
+		return []string{"-xf"}, ".tar", nil
+	case bytes.Equal(header[0:4], []byte{'h', 's', 'q', 's'}):
+		return []string{""}, ".squashfs", nil
+	default:
+		return []string{""}, "", fmt.Errorf("Unsupported compression")
+	}
+
+}
+
+func Unpack(file string, path string, sType StorageType, runningInUserns bool) error {
+	extractArgs, extension, err := DetectCompression(file)
+	if err != nil {
+		return err
+	}
+
+	command := ""
+	args := []string{}
+	if strings.HasPrefix(extension, ".tar") {
+		command = "tar"
+		if runningInUserns {
+			args = append(args, "--wildcards")
+			args = append(args, "--exclude=dev/*")
+			args = append(args, "--exclude=./dev/*")
+			args = append(args, "--exclude=rootfs/dev/*")
+			args = append(args, "--exclude=rootfs/./dev/*")
+		}
+		args = append(args, "-C", path, "--numeric-owner")
+		args = append(args, extractArgs...)
+		args = append(args, file)
+	} else if strings.HasPrefix(extension, ".squashfs") {
+		command = "unsquashfs"
+		args = append(args, "-f", "-d", path, "-n")
+
+		// Limit unsquashfs chunk size to 10% of memory and up to 256MB (default)
+		// When running on a low memory system, also disable multi-processing
+		mem, err := DeviceTotalMemory()
+		mem = mem / 1024 / 1024 / 10
+		if err == nil && mem < 256 {
+			args = append(args, "-da", fmt.Sprintf("%d", mem), "-fr", fmt.Sprintf("%d", mem), "-p", "1")
+		}
+
+		args = append(args, file)
+	} else {
+		return fmt.Errorf("Unsupported image format: %s", extension)
+	}
+
+	output, err := RunCommand(command, args...)
+	if err != nil {
+		// Check if we ran out of space
+		fs := syscall.Statfs_t{}
+
+		err1 := syscall.Statfs(path, &fs)
+		if err1 != nil {
+			return err1
+		}
+
+		// Check if we're running out of space
+		if int64(fs.Bfree) < int64(2*fs.Bsize) {
+			if sType == StorageTypeLvm {
+				return fmt.Errorf("Unable to unpack image, run out of disk space (consider increasing your pool's volume.size).")
```

Comment:
So given that this now applies to more than just LVM and I believe is the reason why you're dragging out the StorageType stuff out of the lxd package (making a huge diff), it's probably easier to just replace the sType argument with a blockBackend (bool) argument and have the warning shown if it's true.

Then in lxd we can call Unpack with blockBackend=true if the backend is LVM or Ceph.

That should significantly reduce the size of this branch.